connect-inject/endpoints_controller_test.go

```diff
@@ -0,0 +1,1322 @@
+				{
+					ServiceID:      "pod1-different-consul-svc-name",
+					ServiceAddress: "1.2.3.4",
+				},
+			},
+			expectedProxySvcInstances: []*api.CatalogService{
+				{
+					ServiceID:      "pod1-different-consul-svc-name-sidecar-proxy",
+					ServiceAddress: "1.2.3.4",
+				},
+			},
+		},
+		{
+			// When a k8s deployment is deleted but it's k8s service continues to exist, the endpoints has no addresses.
+			name:          "Consul has instances that are not in the endpoints, and the endpoints has no addresses.",
+			consulSvcName: "service-updated",
+			k8sObjects: func() []runtime.Object {
+				pod1 := createPod("pod1", "1.2.3.4", true)
+				endpoint := &corev1.Endpoints{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "service-updated",
+						Namespace: "default",
+					},
+				}
+				return []runtime.Object{pod1, endpoint}
+			},
+			initialConsulSvcs: []*api.AgentServiceRegistration{
+				{
+					ID:      "pod1-service-updated",
+					Name:    "service-updated",
+					Port:    80,
+					Address: "1.2.3.4",
+					Meta:    map[string]string{"k8s-service-name": "service-updated", "k8s-namespace": "default"},
+				},
+				{
+					Kind:    api.ServiceKindConnectProxy,
+					ID:      "pod1-service-updated-sidecar-proxy",
+					Name:    "service-updated-sidecar-proxy",
+					Port:    20000,
+					Address: "1.2.3.4",
+					Proxy: &api.AgentServiceConnectProxyConfig{
+						DestinationServiceName: "service-updated",
+						DestinationServiceID:   "pod1-service-updated",
+					},
+					Meta: map[string]string{"k8s-service-name": "service-updated", "k8s-namespace": "default"},
+				},
+				{
+					ID:      "pod2-service-updated",
+					Name:    "service-updated",
+					Port:    80,
+					Address: "2.2.3.4",
+					Meta:    map[string]string{"k8s-service-name": "service-updated", "k8s-namespace": "default"},
+				},
+				{
+					Kind:    api.ServiceKindConnectProxy,
+					ID:      "pod2-service-updated-sidecar-proxy",
+					Name:    "service-updated-sidecar-proxy",
+					Port:    20000,
+					Address: "2.2.3.4",
+					Proxy: &api.AgentServiceConnectProxyConfig{
+						DestinationServiceName: "service-updated",
+						DestinationServiceID:   "pod2-service-updated",
+					},
+					Meta: map[string]string{"k8s-service-name": "service-updated", "k8s-namespace": "default"},
+				},
+			},
+			expectedNumSvcInstances:    0,
+			expectedConsulSvcInstances: []*api.CatalogService{},
+			expectedProxySvcInstances:  []*api.CatalogService{},
+		},
+		{
+			// When a k8s deployment is deleted but it's k8s service continues to exist, the endpoints has no addresses.
+			name:          "Different Consul service name: Consul has instances that are not in the endpoints, and the endpoints has no addresses.",
+			consulSvcName: "different-consul-svc-name",
+			k8sObjects: func() []runtime.Object {
+				pod1 := createPod("pod1", "1.2.3.4", true)
+				pod1.Annotations[annotationService] = "different-consul-svc-name"
```

Comment:
if endpoints doesn't have addresses, then do we need to create a pod?

Reply:
you're totally right!